src/QMCHamiltonians/ECPComponentBuilder_L2.cpp

```diff
@@ -20,7 +20,7 @@ void ECPComponentBuilder::buildL2(xmlNodePtr cur)
   if (grid_global == 0)
   {
     app_error() << "    Global grid needs to be defined." << std::endl;
-    APP_ABORT("ECPComponentBuilder::buildL2");
+    myComm->barrier_and_abort("ECPComponentBuilder::buildL2");
```

Comment:
These seem like they are guaranteed to happen on all nodes so why not throw UniformCommunicateError? I feel like in case of a fatal error the error handling should start with throwing an exception and it should be a UniformCommunicateError if its expected to occur on every rank.

Reply:
The definition of all nodes is encoded in myComm. Once an all-node error reaches a level with accessible communicator, it should be handled rather than keep going up to the top level catch which might have a different definition of all nodes (communicator).

Reply:
Sounds like the scope that owns the myComm should also be the error handling scope.  Doing it this way requires myComm to be known by every bit of code that can be called inside an MPI scope this increases coupling, complicates testing ,and injects access to MPI into tons of classes that don't require it.  Hence some many classes unnecessarily inheriting from MPIObjectBase.

The point of the UniformCommunicateError is to remove this inheritance and therefore dependence on the "message" target from as much of the code as possible.

Reply:
> Sounds like the scope that owns the myComm should also be the error handling scope.
It should. QMCPACK is an MPI program. If the MPI communicator is exposed at a given class, the developer should be aware what operations are serial and what operations are parallel calling from the member function of this class and thus error handling must happen at the level where the communicator is exposed.

It is clearly wrong to catch UniformCommunicateError at the top level. Why we catch UniformCommunicateError at QMCMain.cpp instead of qmcapp.cpp? Because in an ensemble job, the communicator in qmcapp was split into several sub-communicators used by QMCMain. If we catch UniformCommunicateError at qmcapp, the run may hang.

Don't get me wrong. I never say you should not use UniformCommunicateError, If a class doesn't need to be exposed with MPI communicators, the commincator exposure should be removed in the first place and then use UniformCommunicateError to convey a uniform error. However, blindly using UniformCommunicateError is simply dangerous.